### PR TITLE
#34 Mark objects as deleted

### LINKED ISSUE #34: Mark deleted objects
Add a flag to deleted DaabaseObjects that marks them as deleted and causes TDBJ to refuse operations on them

### PR DESCRIPTION
**Pull request by kaklakariada**
Closes #34

### CODE REVIEW

File: src/main/java/com/exasol/dbbuilder/dialects/AbstractUser.java [hunk 1]
```diff
@@ -47,12 +47,12 @@ public String getPassword() {
 
     @Override
     public Map<DatabaseObject, ObjectPrivilege[]> getObjectPrivileges() {
-        return this.objectPrivileges;
+        return Collections.unmodifiableMap(this.objectPrivileges);
```
Comment (ckunki) [inline src/main/java/com/exasol/dbbuilder/dialects/AbstractUser.java:50]: Could we also use `Map.of()`?

File: src/main/java/com/exasol/dbbuilder/dialects/DatabaseObject.java [hunk 2]
```diff
@@ -47,7 +47,8 @@ public interface DatabaseObject {
     boolean isOwned();
 
     /**
-     * Drop the database object and all contained objects.
+     * Drop the database object and all contained objects and mark this as deleted. Other operations will fail when
+     * executing them on a deleted object.
```
Comment (ckunki) [inline src/main/java/com/exasol/dbbuilder/dialects/DatabaseObject.java:51]:
```suggestion
     * Remove this database object from database incl. all contained objects and mark it as deleted. 
     * Later operations on this object will throw a {@link DatabaseObjectDeletedException} Exception.
```

File: src/main/java/com/exasol/dbbuilder/dialects/DatabaseObjectDeletedException.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package com.exasol.dbbuilder.dialects;
+
+/**
+ * This exception is thrown when executing an operation on a {@link DatabaseObject} that was deleted.
```
Comment (ckunki) [inline src/main/java/com/exasol/dbbuilder/dialects/DatabaseObjectDeletedException.java:4]:
```suggestion
 * This exception is thrown when executing an operation on a {@link DatabaseObject} that already had been deleted.
```

File: src/main/java/com/exasol/dbbuilder/dialects/exasol/ConnectionDefinition.java [hunk 4]
```diff
@@ -93,12 +91,16 @@ public boolean hasParent() {
...
+                ExaError.messageBuilder("E-TDBJ-8")
+                        .message("Illegal attempt to access parent object of a CONNECTION which is a top-level object.")
+                        .toString());
     }
 
     @Override
     // [impl->dsn~dropping-connections~1]
-    public void drop() {
+    protected void dropInternally() {
+        verifyNotDeleted();
```
Comment (ckunki) [inline src/main/java/com/exasol/dbbuilder/dialects/exasol/ConnectionDefinition.java:103]: As far as I can understand the super class method `drop()` should already call `verifyNotDeleted()`, right?
```suggestion
    protected void dropInternally() {
```

File: src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolSchema.java [hunk 5]
```diff
@@ -39,6 +37,7 @@ public ExasolSchema(final ExasolImmediateDatabaseObjectWriter writer, final Iden
      */
     public AdapterScript createAdapterScript(final String name, final AdapterScript.Language language,
             final String content) {
+        verifyNotDeleted();
```
Comment (ckunki) [inline src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolSchema.java:40]: Actually the current solution does not convince me to 100%.
It requires changes in many places and I think it is hard to verify, that no place has been forgotten. 
Also in the potential case when creating new types of database objects or when extending the existing classes there is the risk of creating new methods and forgetting to guard them with `verifyNotDeleted()`.

A potential solution that comes into my mind would be to change `AbstractImmediateDatabaseObjectWriter.writeToObject()` and maybe some additional places of that kind. 

On the other hand my proposal could lead to `DatabaseObjectDeletedException` being thrown at a later point in time, which might deviate from the fail-fast-principle. This especially is true in the current case for `AdapterScript.createAdapterScript()` as the method will only return a builder without immediate write operation to the database. 

Also I don't want to block the PR too long. 🙂 

I am happy to discuss.

File: src/test/java/com/exasol/dbbuilder/dialects/TableTest.java [hunk 6]
```diff
@@ -29,38 +31,35 @@ class TableTest {
...
     void testGetType() {
-        assertThat(Table.builder(this.writerMock, this.schemaMock, DummyIdentifier.of("A")).build().getType(),
-                equalTo("table"));
+        assertThat(tableBuilder("A").build().getType(), equalTo("table"));
     }
 
     @Test
     void testGetName() {
-        final Table table = Table.builder(this.writerMock, this.schemaMock, DummyIdentifier.of("FOO")).build();
+        final Table table = tableBuilder("FOO").build();
```
Comment (ckunki) [inline src/test/java/com/exasol/dbbuilder/dialects/TableTest.java:39]: 👍

File: src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleUserTest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    void testGetFullyQualifiedName() {
+        assertThat(new OracleUser(this.writerMock, MySQLIdentifier.of("JOHNDOE")).getFullyQualifiedName(),
+                equalTo("`JOHNDOE`"));
+    }
+
+    @Test
+    void testGetObjectPrivileges(@Mock final DatabaseObject objectMock) {
+        final ObjectPrivilege[] expectedObjectPrivileges = { MySqlObjectPrivilege.INSERT, MySqlObjectPrivilege.DELETE };
+        final User user = new OracleUser(this.writerMock, MySQLIdentifier.of("OBJUSER")) //
+                .grant(objectMock, expectedObjectPrivileges);
```
Comment (ckunki) [inline src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleUserTest.java:46]: Is it OK to use MySqlObjectPrivilege here? 
Maybe we could add an alias with "Oracle" prefix? 
```suggestion
        final ObjectPrivilege[] expectedObjectPrivileges = { MySqlObjectPrivilege.INSERT, MySqlObjectPrivilege.DELETE };
        final User user = new OracleUser(this.writerMock, OracleIdentifier.of("OBJUSER")) //
                .grant(objectMock, expectedObjectPrivileges);
```

**Review by ckunki — CHANGES_REQUESTED** [review]

File: src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleUserTest.java (hunk 7 above)
Comment (kaklakariada) [reply src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleUserTest.java:46]: Ooops... Thanks for the hint :)

File: src/main/java/com/exasol/dbbuilder/dialects/AbstractUser.java (hunk 1 above)
Comment (kaklakariada) [reply src/main/java/com/exasol/dbbuilder/dialects/AbstractUser.java:50]: No, it only allows creating maps from keys & values.
Plus this makes the intention of an unmodifiable map explicit.

**Review by ckunki — APPROVED** [review]

**Review by ckunki — APPROVED** [review]

**Pull request merged**